### DISCUSSION TITLE
Can the ergonomics of serde+structures be made better than in Rusoto?

### DISCUSSION THREAD

**Discussion by pcn** [awslabs/aws-sdk-rust#78] [General]
I've taken a bit of a dive into the rabbit hole of trying to use some of the structures generated for rusoto-ec2. https://github.com/rusoto/rusoto/issues/1826 is the summary of where I gave up. It seems like the complexity of representing all of the possible ec2 options has made the resulting structures impossible for serde to make sense of.

I think that being able to make the various structures serde-derivable should be an explicit goal if at all possible to make accessing the various API results much easier going forward.

**Comment by rcoh** [comment]
Yeah this is an interesting issue that's been on my radar. I'd like to really understand the use case for serializing structures outside of sending requests to AWS. (Eg: if the use case is reading and writing operations to disk, it may make sense for us to support that directly, rather than going through serde)

 A couple of other notes:
- Although _some_ structs currently implement `Serialize`/`Deserialize` they won't in the long term (we don't want to expose this implementation detail and we may want to change the underlying serialization in the future)
- I'm considering if we should we add `serde derives` for structures behind a `cfg` flag but even that may be limiting in terms of potential API evolution

**Reply by pcn** [reply]
My goal is to be able to get relevant bits of an EC2 instance, RDS instances, etc. - interfaces, tags, disks, etc. which can vary wildly in terms of their count, existence, valid properties etc. in a way that can be flexible at run-time since these things are all over the place.

The tool I've got that I'm trying to replace is my first useful rust tool, and it's aged badly. E.g. this approach doesn't work for nodes that add and remove interfaces a lot, as eks worker nodes do. 

I'm looking for a way to get this sort of info without the corners that I've painted myself into there. 

Edit: here's the structure I'm using currently, which is what I'm hoping to improve on: https://github.com/pcn/raal/blob/master/src/lib.rs#L107

**Reply by rpadovani** [reply]
Hi! I chime in to add my use case, hope it can be helpful.
Side note: it is not for an important project, it is something I am doing on my PC to learn a bit of Rust.

Basically, I am using [Tauri](https://tauri.studio) to create the GUI for an application: the backend of the application is in Rust, and I'd like to send messages between the backend and the frontend. To do so, Tauri can pass data of any type, as long as they implement Serde::Serialize https://tauri.studio/en/docs/usage/guides/command#returning-data.

It would be easier for me to pass directly the result of an invocation of the SDK, instead of having to declare my own struct.

As I said, however, I am learning Rust, so maybe I'm missing something here.

Thanks!

**Comment by jmklix** [comment]
Created feature request [here](https://github.com/awslabs/aws-sdk-rust/issues/507) and locked discussion
